### ISSUE TITLE
rolling updates for AMIs in EKS worker autoscaling group

### ISSUE THREAD

**Issue by danielhartnell** [mozilla-iam/eks-deployment#79]
I did not plan for this when I created the cluster. We need a good process for cycling out the EKS worker AMIs without introducing downtime. Ideally, Terraform could just replace the instances one at a time in the autoscaling group while retaining n instances at any given time. Given adequate time between instance termination, the Kubernetes control plane should have time to move pods to the new instances.

At this point, I think we can either create a new autoscaling group and wait for those instances to be healthy before we destroy the old autoscaling group. This is a nice way to switch over to new nodes but since we have to actually move workloads from old nodes to new nodes, I don't love it.

Alternatively, I like the idea of using AutoScalingRollingUpdate settings in CloudFormation. This means I'd have to use Terraform to create and apply a CloudFormation stack for the autoscaling group.

In either case, I'll try to see if we can leverage health checks to do this as safely as we can. Once it's done, I want to update the EKS-optimized AMI that we're using and increase the root volume size.

**Comment by danielhartnell** [comment]
Great idea here: https://thinklumo.com/blue-green-node-deployment-kubernetes-eks-terraform/

**Comment by The-smooth-operator** [comment]
We did implement the idea described in the post above.
First time we tried the rolling update was a bit messy, but thanks to some tweaks Daniel did in the ASG worker groups, now is possible to perform updates in the AMI of the EC2 instances without downtime.

The process is still very manual:
 - Modify the AMI in the green worker group
 - Increase from 0 to N the number of instances on the Green group. Terraform apply
 - Once they join the cluster, kubectl drain the old nodes, one by one
 - After all nodes are drained and pods are only running in Green worker group, we set the number of instances in Blue ASG group to 0. Terraform apply.
 - Finally, we have to remove manually the EC2 instances as the ASG is not doing it.

While we can automate this via custom scripts, the process looks solid, and I'd consider it done.
